# **You're A Winner, Baby!**

**Author:** Anton Reyes

## **Introduction**

### **Requirements and Imports**

#### **Imports**

**Basic Libraries**

* `numpy` contains a large collection of mathematical functions
* `pandas` contains functions that are designed for data manipulation and data analysis

In [91]:
import numpy as np
import pandas as pd

**Visualization Libraries**

* `matplotlib.pyplot` contains functions to create interactive plots
* `seaborn` is a library based on matplotlib that allows for data visualization
* `plotly` is an open-source graphing library for Python.

In [92]:
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

**Natural Language Processing Libraries**
* `re` is a module that allows the use of regular expressions

In [93]:
import re

#### **Datasets and Files**

The following `csv` file was used for this project:

- `finalists.csv` contains all the finalists of the Drag Race franchise as well as the placements of each contestant in their seasons. Finalists are as of April 22, 2023

## **Data Collection**

Importing the dataset using pandas.

In [94]:
dataset = "data/finalists.csv"

df = pd.read_csv(dataset)
df.head()

Rank WRE   Rank Country Season    Code W/R/E          Queen  Z-Score  \
0       NaN    NaN     NaN    NaN     NaN   NaN            NaN      NaN   
1      18.0    8.0      US    AS1  US_AS1     W  Chad Michaels     0.47   
2     131.0  159.0      US    AS1  US_AS1     R          Raven    -2.26   
3      77.0  159.0      US    AS1  US_AS1     E        Jujubee    -2.26   
4       6.0    8.0      US    AS1  US_AS1     E        Shannel     0.47   

    Percent  Score  ...  Low  Safe  High  SWin  Win  LSA    T  Unnamed: 36  \
0       NaN    NaN  ...   -1     0     1   1.5    2  NaN  NaN     Never/--   
1   100.00%    5.0  ...    0     0     1   0.0    3  NaN  NaN           --   
2  -120.00%   -6.0  ...    1     0     1   0.0    0  LSA  2.0           --   
3  -120.00%   -6.0  ...    1     0     1   0.0    0  LSA  2.0           --   
4   100.00%    5.0  ...    0     0     1   0.0    3  NaN  NaN           --   

   Unnamed: 37  Unnamed: 38  
0         Rank         Rank  
1        1,220            1  
2          237          140  
3          237          140  
4        1,220            1  

[5 rows x 39 columns]

## **Description of the Dataset**

Here, we find the shape of the dataset.

In [95]:
df.shape

(161, 39)

By looking at the `info` of the dataframe, we can see that there are `non-null` values. 

In [96]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 161 entries, 0 to 160
Data columns (total 39 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Rank WRE     160 non-null    float64
 1   Rank         160 non-null    float64
 2   Country      160 non-null    object 
 3   Season       160 non-null    object 
 4   Code         160 non-null    object 
 5   W/R/E        160 non-null    object 
 6   Queen        160 non-null    object 
 7   Z-Score      160 non-null    float64
 8   Percent      160 non-null    object 
 9   Score        160 non-null    float64
 10  Episode      160 non-null    float64
 11  1            154 non-null    float64
 12  2            154 non-null    float64
 13  3            156 non-null    float64
 14  4            160 non-null    float64
 15  5            159 non-null    float64
 16  6            142 non-null    float64
 17  7            139 non-null    float64
 18  8            93 non-null     float64
 19  9       

The dataset itself contains the progress of all contestants in all the different franchises of Drag Race. That being said, the original file (in Google Sheets) contains more than 10 tabs for the different franchises, seasons, and countries - and it's still growing. 

Manually, I take the data/progess for each contestant that reach the final episode - the finalists. 

As we can see from the `Queen` column, there are 160 finalists regardless of season. This is because there are repeating finalists due to All-Star seasons. 

## **Exploratory Data Analysis Part 1**

The following questions are asked to guide the EDA.

1. How many finalists are there per class?
2. How many unique values are there in the `Season` column?
3. How many unique values are there in the `Country` column?
4. What are measures of central tendency in the `Episode` column?

### **1. How many finalists are there per class?**

Before getting the number of contestants per class, we first define the classes:

| Class | Meaning | Definition |
|:---:|---| --- |
| W | Winner | Contestant has won the season |
| R | Runner-Up | Contestant became the runner-up or placed 2nd |
| E | Elimenated | Contestant was elimenated at the finale and placed 3rd/4th |

With that, we now get the total of contestants by counting them in the `W/R/E` column.

In [97]:
df[["W/R/E"]].count()

W/R/E    160
dtype: int64

Now, we get the division of classes between all 160 contestants by printing the `.value_counts()` and summing it all up to measure.

In [98]:
print("Division of classes:", df[["W/R/E"]].value_counts())
print("Sum of the classes:", df[["W/R/E"]].value_counts().sum())

Division of classes: W/R/E
R        69
W        46
E        45
dtype: int64
Sum of the classes: 160


### **2. How many unique values are there in the `Season` column?**

Similar to the previous question, we first get the count of seasons available.

In [99]:
df[['Season']].count()

Season    160
dtype: int64

From getting the value counts in the `Season` column, we can see that there are inconsistensies in the dataset. Here, we can take note of how to format the dataset when it comes to preprocessing the dataframe.

In [100]:
df[['Season']].value_counts()

Season
1         26
S1        24
2         16
S2        10
3          7
14         5
15         4
AS6        4
S13        4
S11        4
AS7        4
S10        4
AS4        4
AS3        4
AS2        4
AS1        4
4          4
S9         4
AS5        3
S12        3
S3         3
S4         3
S5         3
S6         3
S7         3
S8         3
dtype: int64

### **3. How many unique values are there in the `Country` column?**

In [101]:
df[["Country"]].count()

Country    160
dtype: int64

By seeing the value counts in the `Country` column, we can observer that there are multiple values. However, countries with `_WORLD` just classify them as a different country rather then a *type* of season. When in fact, `_WORLD` seasons are similar to an All Star season.

In [102]:
df[["Country"]].value_counts()

Country  
US           79
UK           14
CAN          10
ESP           8
AUS           7
HOL           7
ITA           7
THI           6
CAN_WORLD     4
FRA           4
PH            4
UK_WORLD      4
BEL           3
SWE           3
dtype: int64

### **4. What are the measure of central tendency in the `Episode` column?**

First, we check the data type in the `Episode` column.

In [103]:
df['Episode'].dtypes

dtype('float64')

Since the values in the episode column are floats, we can `describe` the column to get the inferential statistics.

In [104]:
df["Episode"].describe()

count    160.00000
mean       8.81875
std        2.40511
min        5.00000
25%        7.00000
50%        9.00000
75%       11.00000
max       13.00000
Name: Episode, dtype: float64

In [105]:
print("Max:", df['Episode'].max()),
print("Min:", df['Episode'].min()),
print("Range:", df['Episode'].max() - df['Episode'].min())

Max: 13.0
Min: 5.0
Range: 8.0


## **Data Preprocessing**

#### **Data Preprocessing**

##### **Dropping Unneeded Columns**

This dataset has been worked on in a Google Sheet in order to record multiple franchises at once. That being said, the analysis itself has been done in the same sheet. This resulted into a lot of columns with formulas rather than actual data.

In [106]:
df.head()

Rank WRE   Rank Country Season    Code W/R/E          Queen  Z-Score  \
0       NaN    NaN     NaN    NaN     NaN   NaN            NaN      NaN   
1      18.0    8.0      US    AS1  US_AS1     W  Chad Michaels     0.47   
2     131.0  159.0      US    AS1  US_AS1     R          Raven    -2.26   
3      77.0  159.0      US    AS1  US_AS1     E        Jujubee    -2.26   
4       6.0    8.0      US    AS1  US_AS1     E        Shannel     0.47   

    Percent  Score  ...  Low  Safe  High  SWin  Win  LSA    T  Unnamed: 36  \
0       NaN    NaN  ...   -1     0     1   1.5    2  NaN  NaN     Never/--   
1   100.00%    5.0  ...    0     0     1   0.0    3  NaN  NaN           --   
2  -120.00%   -6.0  ...    1     0     1   0.0    0  LSA  2.0           --   
3  -120.00%   -6.0  ...    1     0     1   0.0    0  LSA  2.0           --   
4   100.00%    5.0  ...    0     0     1   0.0    3  NaN  NaN           --   

   Unnamed: 37  Unnamed: 38  
0         Rank         Rank  
1        1,220            1  
2          237          140  
3          237          140  
4        1,220            1  

[5 rows x 39 columns]

Using pandas, we get all the columns in the dataframe into a list for easier reading.

In [107]:
df.columns

Index(['Rank WRE', 'Rank', 'Country', 'Season', 'Code', 'W/R/E', 'Queen',
       'Z-Score', 'Percent', 'Score', 'Episode', '1', '2', '3', '4', '5', '6',
       '7', '8', '9', '10', '11', '12', '13', '14', '15', 'Mean: 3.11',
       'Unnamed: 27', 'Bttm', 'Low', 'Safe', 'High', 'SWin', 'Win', 'LSA', 'T',
       'Unnamed: 36', 'Unnamed: 37', 'Unnamed: 38'],
      dtype='object')

After getting the list of columns, we now make a list of columns to drop.

The columns below are the columns that were done using formulas in Google Sheets. 

In [108]:
drop_col = ['Rank WRE', 'Rank', 'Code', 'Episode', 'Percent', 'Score', 
            'Mean: 3.11', 'Z-Score', 'Unnamed: 27', 'LSA', 'T', 
            'Unnamed: 36', 'Unnamed: 37', 'Unnamed: 38']

Because of an extra row to further divide the original dataset in Google Sheets, we drop the first row since it won't be usable in any way.

In [109]:
df = df.drop(drop_col, axis=1) #Dropping the list of columns
df = df.iloc[1:,:] #Dropping the first row

In [110]:
df.head() #checking the dataset

Country Season W/R/E          Queen    1    2    3    4    5    6  ...  12  \
1      US    AS1     W  Chad Michaels -2.0  1.0  2.0  2.0  2.0  NaN  ... NaN   
2      US    AS1     R          Raven  1.0 -1.0 -2.0 -2.0 -2.0  NaN  ... NaN   
3      US    AS1     E        Jujubee  1.0 -1.0 -2.0 -2.0 -2.0  NaN  ... NaN   
4      US    AS1     E        Shannel -2.0  1.0  2.0  2.0  2.0  NaN  ... NaN   
5      US    AS2     W         Alaska  1.0  2.0  0.0  2.0  2.0  2.0  ... NaN   

   13  14  15  Bttm  Low  Safe  High  SWin  Win  
1 NaN NaN NaN     1    0     0     1   0.0    3  
2 NaN NaN NaN     3    1     0     1   0.0    0  
3 NaN NaN NaN     3    1     0     1   0.0    0  
4 NaN NaN NaN     1    0     0     1   0.0    3  
5 NaN NaN NaN     1    0     1     1   0.0    4  

[5 rows x 25 columns]

##### **Regex Functions**

Because on inconsistencies in the `Country` and `Season` columns, we make functions using RegEx to do so. Because the functions will be removing certain characters and strings, another column will be created later on to help specify certain seasons.

Other functions will be used later on.

In [111]:
#regex function to remove S but not remove the S in AS
def remove_s(x):
    return re.sub(r'S', '', x)

#regex function to remove AS and _WORLD
def remove_as(x):
    return re.sub(r'AS|_WORLD', '', x)

#regex function to remove the letter A
def remove_a(x):
    return re.sub(r'A', '', x)


##### **Removing `S` from the Season column**

Here, we apply the `remove_s` function to remove seasons that have the letter `S` even though the number already indicated the season.

In [112]:
df['Season'] = df['Season'].apply(remove_s)
df['Season'].value_counts()

1     50
2     26
3     10
4      7
14     5
A7     4
A6     4
9      4
A2     4
A1     4
13     4
11     4
10     4
A4     4
A3     4
15     4
12     3
5      3
6      3
7      3
8      3
A5     3
Name: Season, dtype: int64

With this new column, the S has been removed and only the All Stars Seasons are identified

##### **Regex Functions for New Column**

After removing the `S` in the `Season` column, we now make functions to extract certain strings to a new column to identify the season if it was a Regular, All Star, or World season.

In [114]:
def identify_world(x):
    if '_WORLD' in x:
        return 'WORLD',
    
def identify_as(x):    
    if 'A' in x:
        return 'AS'

def identify_reg(x):
    if 'A' not in x:
        return 'REG'
    
def remove_world_reg(df, column_name):
    df[column_name] = df[column_name].str.replace(r"\('WORLD',\)REG", "WORLD", regex=True)
    return df
    


##### **Extracting the formats**

By using the functions, we now apply it to their respective columns. We also replace `None` values as empty strings so that when it comes to joining the three columns, there would be no problem.

In [115]:
df['world'] = df['Country'].apply(identify_world).astype(str)
df['world'] = df['world'].replace('None', '')
df['world'].value_counts()

              152
('WORLD',)      8
Name: world, dtype: int64

In [116]:
df['allstars'] = df['Season'].apply(identify_as).astype(str)
df['allstars'] = df['allstars'].replace('None', '')
df['allstars'].value_counts()

      133
AS     27
Name: allstars, dtype: int64

In [117]:
df['regular'] = df['Season'].apply(identify_reg).astype(str)
df['regular'] = df['regular'].replace('None', '')
df['regular'].value_counts()

REG    133
        27
Name: regular, dtype: int64

In [118]:
#joining two columns into one column
df['format'] = df['world'] + df['allstars'] + df['regular']

df['format'] = df['format'].replace(np.nan, 'REG')

df['format'].value_counts()

REG              125
AS                27
('WORLD',)REG      8
Name: format, dtype: int64

After joining all the new columns, we can see that the column for World seasons show `('WORLD',)REG` instead of `WORLD`. This will later be on cleaned to fit the format needed to make it readable and understandable at a glance.

#### **Data Cleaning**

##### **Fixing the Columns**

After extracting all the needed information, we can now clean the columns accordingly since the functions rely on certain strings to function.

By using the previously made RegEx functions we now remove the letter `A` in the `Season` column.

In [119]:
df['Season'] = df['Season'].apply(remove_a)
df['Season'].value_counts()

1     54
2     30
3     14
4     11
6      7
7      7
5      6
14     5
10     4
11     4
13     4
9      4
15     4
12     3
8      3
Name: Season, dtype: int64

Similar to the previous cell, we now fix the format of the `format` column for those with the `('WORLD',)REG` to `WORLD` instead.

In [137]:
df = remove_world_reg(df, 'format')
df['format'].value_counts()

REG      125
AS        27
WORLD      8
Name: format, dtype: int64

##### **Removing Type in Season**

We now remove the string `_WORLD` column since we've already identified in another column which season and country is a `WORLD` season.

In [121]:
df['Country'] = df['Country'].apply(remove_as)
df['Country'].value_counts()

US     79
UK     18
CAN    14
ESP     8
HOL     7
AUS     7
ITA     7
THI     6
FRA     4
PH      4
BEL     3
SWE     3
Name: Country, dtype: int64

##### **Changing Placement Values**

In the original dataset, values included negative values; Specifically `-2` and `-1` which indicated Bottom 2 and Low placements, respectively. In order to have a better 'view' of the scores and placements, we first make a lsit of episodes.

In [123]:
episodes =[ '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14']

After making the list, we now replace the values in the `episodes` columns using the pandas function `replace` as well as using a dictionary.

The following placement values will be replaced as follows:

|Old Value|New Value|Placement|Meaning|
|:---:|:---:|---|---|
|-2|0|Bottom 2|Was up for elimination|
|-1|1|Low|Received negative critqiues, but not up for elimination|
|0|2|Safe|Not high or low|
|1|3|High|Received positive critiques, but did not win the challenge or in select seasons, not part of the Top 2|
|1.5|4|Top 2|Top 2 contestants of the week, but lost the lipsync. Applicable to All Starts, World, and selected episodes in regular seasons|
|2|5|Win|Won the main challenge or in some seasons/episodes, won the Top 2 lipsync|

In [124]:
#replacing values with another value in multiple columns
df[episodes] = df[episodes].replace({
    -2 : 0, #Bottom 2 placement
    -1 : 1, #Received negative critiques, but was not up for elimination   
    0 : 2, #Safe 
    1 : 3, #Received positive critiques, but did not win the challenge 
    1.5 : 4, #Top 2 contestants (Applicable in All Stars and World seasons and selected episodes in regular seasons) but lost the lipsync
    2 : 5 #Won the main challenge
    })

Here, we used the `.describe()` function to see the inferential statistics of each episode.

In [125]:
df[episodes].describe()

1           2           3           4           5           6  \
count  154.000000  154.000000  156.000000  160.000000  159.000000  142.000000   
mean     2.629870    2.538961    2.493590    2.646875    2.698113    2.549296   
std      1.231139    1.473660    1.452459    1.618094    1.697931    1.578119   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      2.000000    2.000000    2.000000    2.000000    2.000000    2.000000   
50%      2.000000    2.000000    2.000000    2.000000    3.000000    3.000000   
75%      3.000000    3.000000    3.000000    4.000000    5.000000    3.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

                7          8          9         10         11         12  \
count  139.000000  93.000000  89.000000  51.000000  54.000000  43.000000   
mean     2.438849   2.591398   2.617978   2.411765   2.833333   2.255814   
std      1.711492   1.764607   1.805960   1.779623   1.701830   1.329035   
min      0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   
25%      1.000000   1.000000   1.000000   1.000000   2.000000   2.000000   
50%      2.000000   3.000000   3.000000   3.000000   3.000000   2.000000   
75%      3.000000   4.000000   5.000000   3.500000   4.750000   3.000000   
max      5.000000   5.000000   5.000000   5.000000   5.000000   5.000000   

              13        14  
count  13.000000  9.000000  
mean    2.461538  1.888889  
std     1.941451  1.536591  
min     0.000000  0.000000  
25%     0.000000  1.000000  
50%     3.000000  2.000000  
75%     3.000000  2.000000  
max     5.000000  5.000000

##### **Sum of placement scores**

Because of the previously dropped score column we now make a new one with the new values that were just replaced. We sum the episode columns and add it to a new column.

In [126]:
#sum all numbers per row in columns 1-14
df['score'] = df[episodes].sum(axis=1)

df

Country Season W/R/E                Queen    1    2    3    4    5    6  \
1        US      1     W        Chad Michaels  0.0  3.0  5.0  5.0  5.0  NaN   
2        US      1     R                Raven  3.0  1.0  0.0  0.0  0.0  NaN   
3        US      1     E              Jujubee  3.0  1.0  0.0  0.0  0.0  NaN   
4        US      1     E              Shannel  0.0  3.0  5.0  5.0  5.0  NaN   
5        US      2     W               Alaska  3.0  5.0  2.0  5.0  5.0  5.0   
..      ...    ...   ...                  ...  ...  ...  ...  ...  ...  ...   
156     BEL      1     W         Drag Couenne  3.0  5.0  2.0  5.0  2.0  3.0   
157     BEL      1     E                Susan  2.0  1.0  5.0  0.0  3.0  1.0   
158     SWE      1     W  Admira Thunderpussy  3.0  2.0  2.0  5.0  5.0  1.0   
159     SWE      1     R              Fontana  2.0  2.0  1.0  2.0  0.0  5.0   
160     SWE      1     E          Vanity Vain  2.0  0.0  5.0  1.0  3.0  3.0   

     ...  Low  Safe  High  SWin  Win  world  allstars  regular  format  score  
1    ...    0     0     1   0.0    3               AS               AS   18.0  
2    ...    1     0     1   0.0    0               AS               AS    4.0  
3    ...    1     0     1   0.0    0               AS               AS    4.0  
4    ...    0     0     1   0.0    3               AS               AS   18.0  
5    ...    0     1     1   0.0    4               AS               AS   25.0  
..   ...  ...   ...   ...   ...  ...    ...       ...      ...     ...    ...  
156  ...    0     2     2   0.0    3                       REG     REG   25.0  
157  ...    3     1     1   0.0    1                       REG     REG   13.0  
158  ...    1     2     1   0.0    3                       REG     REG   23.0  
159  ...    1     4     0   0.0    1                       REG     REG   14.0  
160  ...    1     1     2   0.0    1                       REG     REG   14.0  

[160 rows x 30 columns]

##### **Placement Count**

Here, we count the number of placements for each contestants in the `episodes` columns. 

For each placement, there is a designated column.

In [127]:
#count the number of 0, 1, 2, 3, 4, 5 in each row per contestant
df['bottom'] = df[episodes].eq(0).sum(axis=1)
df['low'] = df[episodes].eq(1).sum(axis=1)
df['safe'] = df[episodes].eq(2).sum(axis=1)
df['high'] = df[episodes].eq(3).sum(axis=1)
df['semiwin'] = df[episodes].eq(4).sum(axis=1)
df['win'] = df[episodes].eq(5).sum(axis=1)

In [128]:
placement = ['bottom', 'low', 'safe', 'high', 'semiwin', 'win']

df

Country Season W/R/E                Queen    1    2    3    4    5    6  \
1        US      1     W        Chad Michaels  0.0  3.0  5.0  5.0  5.0  NaN   
2        US      1     R                Raven  3.0  1.0  0.0  0.0  0.0  NaN   
3        US      1     E              Jujubee  3.0  1.0  0.0  0.0  0.0  NaN   
4        US      1     E              Shannel  0.0  3.0  5.0  5.0  5.0  NaN   
5        US      2     W               Alaska  3.0  5.0  2.0  5.0  5.0  5.0   
..      ...    ...   ...                  ...  ...  ...  ...  ...  ...  ...   
156     BEL      1     W         Drag Couenne  3.0  5.0  2.0  5.0  2.0  3.0   
157     BEL      1     E                Susan  2.0  1.0  5.0  0.0  3.0  1.0   
158     SWE      1     W  Admira Thunderpussy  3.0  2.0  2.0  5.0  5.0  1.0   
159     SWE      1     R              Fontana  2.0  2.0  1.0  2.0  0.0  5.0   
160     SWE      1     E          Vanity Vain  2.0  0.0  5.0  1.0  3.0  3.0   

     ...  allstars  regular  format  score  bottom  low  safe  high  semiwin  \
1    ...        AS               AS   18.0       1    0     0     1        0   
2    ...        AS               AS    4.0       3    1     0     1        0   
3    ...        AS               AS    4.0       3    1     0     1        0   
4    ...        AS               AS   18.0       1    0     0     1        0   
5    ...        AS               AS   25.0       1    0     1     1        0   
..   ...       ...      ...     ...    ...     ...  ...   ...   ...      ...   
156  ...                REG     REG   25.0       0    0     2     2        0   
157  ...                REG     REG   13.0       1    3     1     1        0   
158  ...                REG     REG   23.0       0    1     2     1        0   
159  ...                REG     REG   14.0       1    1     4     0        0   
160  ...                REG     REG   14.0       2    1     1     2        0   

     win  
1      3  
2      0  
3      0  
4      3  
5      4  
..   ...  
156    3  
157    1  
158    3  
159    1  
160    1  

[160 rows x 36 columns]

##### **Final Column Fixes**

In preparation for the another set of EDA, we must prepare the dataframe in a better order. We first get all the columns in their current order:

In [129]:
df.columns

Index(['Country', 'Season', 'W/R/E', 'Queen', '1', '2', '3', '4', '5', '6',
       '7', '8', '9', '10', '11', '12', '13', '14', '15', 'Bttm', 'Low',
       'Safe', 'High', 'SWin', 'Win', 'world', 'allstars', 'regular', 'format',
       'score', 'bottom', 'low', 'safe', 'high', 'semiwin', 'win'],
      dtype='object')

Because of the new adjustments, we have to drop columns again before anything else so that we can only see the usefeul features.

In [130]:
drop_col = ['15', 'world', 'allstars', 'regular']
df = df.drop(drop_col, axis=1)
df

Country Season W/R/E                Queen    1    2    3    4    5    6  \
1        US      1     W        Chad Michaels  0.0  3.0  5.0  5.0  5.0  NaN   
2        US      1     R                Raven  3.0  1.0  0.0  0.0  0.0  NaN   
3        US      1     E              Jujubee  3.0  1.0  0.0  0.0  0.0  NaN   
4        US      1     E              Shannel  0.0  3.0  5.0  5.0  5.0  NaN   
5        US      2     W               Alaska  3.0  5.0  2.0  5.0  5.0  5.0   
..      ...    ...   ...                  ...  ...  ...  ...  ...  ...  ...   
156     BEL      1     W         Drag Couenne  3.0  5.0  2.0  5.0  2.0  3.0   
157     BEL      1     E                Susan  2.0  1.0  5.0  0.0  3.0  1.0   
158     SWE      1     W  Admira Thunderpussy  3.0  2.0  2.0  5.0  5.0  1.0   
159     SWE      1     R              Fontana  2.0  2.0  1.0  2.0  0.0  5.0   
160     SWE      1     E          Vanity Vain  2.0  0.0  5.0  1.0  3.0  3.0   

     ...  SWin  Win  format  score  bottom  low  safe  high  semiwin  win  
1    ...   0.0    3      AS   18.0       1    0     0     1        0    3  
2    ...   0.0    0      AS    4.0       3    1     0     1        0    0  
3    ...   0.0    0      AS    4.0       3    1     0     1        0    0  
4    ...   0.0    3      AS   18.0       1    0     0     1        0    3  
5    ...   0.0    4      AS   25.0       1    0     1     1        0    4  
..   ...   ...  ...     ...    ...     ...  ...   ...   ...      ...  ...  
156  ...   0.0    3     REG   25.0       0    0     2     2        0    3  
157  ...   0.0    1     REG   13.0       1    3     1     1        0    1  
158  ...   0.0    3     REG   23.0       0    1     2     1        0    3  
159  ...   0.0    1     REG   14.0       1    1     4     0        0    1  
160  ...   0.0    1     REG   14.0       2    1     1     2        0    1  

[160 rows x 32 columns]

For better calling, we now make all the columns lowercase through mapping.

In [131]:
#lowercase all column names
df.columns = map(str.lower, df.columns)

#### **Feature Selection**

In [138]:
df.columns

Index(['country', 'season', 'w/r/e', 'queen', '1', '2', '3', '4', '5', '6',
       '7', '8', '9', '10', '11', '12', '13', '14', 'bttm', 'low', 'safe',
       'high', 'swin', 'win', 'format', 'score', 'bottom', 'low', 'safe',
       'high', 'semiwin', 'win'],
      dtype='object')

**Select and reordered columns**

To better organize certain data for their own uses, we separate the dataframes based on certain features

In [132]:
df_scores = df[['season', 'country', 'format', 'w/r/e', 'queen', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14']]

df_placements = df[['season', 'country', 'format', 'w/r/e', 'queen', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14']]

## **Exploratory Data Analysis Part 2**

The following questions are asked to guide the EDA.

1. 

In [133]:
df

country season w/r/e                queen    1    2    3    4    5    6  \
1        US      1     W        Chad Michaels  0.0  3.0  5.0  5.0  5.0  NaN   
2        US      1     R                Raven  3.0  1.0  0.0  0.0  0.0  NaN   
3        US      1     E              Jujubee  3.0  1.0  0.0  0.0  0.0  NaN   
4        US      1     E              Shannel  0.0  3.0  5.0  5.0  5.0  NaN   
5        US      2     W               Alaska  3.0  5.0  2.0  5.0  5.0  5.0   
..      ...    ...   ...                  ...  ...  ...  ...  ...  ...  ...   
156     BEL      1     W         Drag Couenne  3.0  5.0  2.0  5.0  2.0  3.0   
157     BEL      1     E                Susan  2.0  1.0  5.0  0.0  3.0  1.0   
158     SWE      1     W  Admira Thunderpussy  3.0  2.0  2.0  5.0  5.0  1.0   
159     SWE      1     R              Fontana  2.0  2.0  1.0  2.0  0.0  5.0   
160     SWE      1     E          Vanity Vain  2.0  0.0  5.0  1.0  3.0  3.0   

     ...  swin  win  format  score  bottom  low  safe  high  semiwin  win  
1    ...   0.0    3      AS   18.0       1    0     0     1        0    3  
2    ...   0.0    0      AS    4.0       3    1     0     1        0    0  
3    ...   0.0    0      AS    4.0       3    1     0     1        0    0  
4    ...   0.0    3      AS   18.0       1    0     0     1        0    3  
5    ...   0.0    4      AS   25.0       1    0     1     1        0    4  
..   ...   ...  ...     ...    ...     ...  ...   ...   ...      ...  ...  
156  ...   0.0    3     REG   25.0       0    0     2     2        0    3  
157  ...   0.0    1     REG   13.0       1    3     1     1        0    1  
158  ...   0.0    3     REG   23.0       0    1     2     1        0    3  
159  ...   0.0    1     REG   14.0       1    1     4     0        0    1  
160  ...   0.0    1     REG   14.0       2    1     1     2        0    1  

[160 rows x 32 columns]

# **Saving Dataframes as CSVs**

In [134]:
#main_df.to_csv('data\main.csv')
